src/lib.rs

```diff
@@ -351,6 +350,38 @@ impl Formatter for CanonicalFormatter {
     }
 }
 
+pub trait CanonJsonSerialize {
+    fn to_canon_json_writer<W>(&self, writer: W) -> Result<()>
+    where
+        W: Write;
+    fn to_canon_json_vec(&self) -> Result<Vec<u8>>;
+    fn to_canon_json_string(&self) -> Result<String>;
+}
+
+impl<S> CanonJsonSerialize for S
+where
+    S: Serialize,
+{
+    fn to_canon_json_writer<W>(&self, writer: W) -> Result<()>
+    where
+        W: Write,
+    {
+        let mut ser = serde_json::Serializer::with_formatter(writer, CanonicalFormatter::new());
+        Ok(self.serialize(&mut ser)?)
+    }
+
+    fn to_canon_json_vec(&self) -> Result<Vec<u8>> {
+        let mut buf = Vec::new();
+        self.to_canon_json_writer(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn to_canon_json_string(&self) -> Result<String> {
+        // We do not emit invalid UTF-8.
+        Ok(unsafe { String::from_utf8_unchecked(self.to_canon_json_vec()?) })
```

Comment:
Presumably you're basing this on https://github.com/serde-rs/json/blob/cd55b5a0ff5f88f1aeb7a77c1befc9ddb3205201/src/ser.rs#L2243 ?

I mean yes, I agree we shouldn't and almost certainly aren't, but I am not quite sure I am comfortable with lifting the `#[forbid(unsafe_code)]` just for this. I mean, the use cases for `to_string` are almost always not seriously performance sensitive anyways.

So my preference here is keep the `#[forbid(unsafe_code)]` and just use `String::from_utf8`.

Reply:
It's also because this is io::Error right now, but I can switch to anyhow::Error

Reply:
Done